### PR TITLE
Add test utilities for comparing expressions

### PR DESCRIPTION
**Pull request by jsiirola**
## Summary/Motivation:
This PR defines two new functions for comparing expression trees in unittest test suites: `assertExpressionsEqual()` and `assertExpressionsStructurallyEqual()`  These functions support more robust comparison of Expression trees using a convenient syntax like:
```python
    def test_simpleSum_API(self):
        m = ConcreteModel()
        m.a = Var()
        m.b = Var()
        e = m.a + m.b
        e += (2*m.a)
        assertExpressionsEqual(
            self,
            e,
            LinearExpression([
                MonomialTermExpression((1, m.a)),
                MonomialTermExpression((1, m.b)),
                MonomialTermExpression((2, m.a)),
            ])
        )
```

## Changes proposed in this PR:
- add  `assertExpressionsEqual()` and `assertExpressionsStructurallyEqual()`
- add tests exercising the test functions

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/core/expr/compare.py [hunk 1]
```diff
@@ -185,3 +189,79 @@ def compare_expressions(expr1, expr2, include_named_exprs=True):
...
+        test.fail(f"Expressions not equal:\n\t"
+                  f"{tostr(prefix_a)}\n\t!=\n\t{tostr(prefix_b)}")
+
+
+def assertExpressionsStructurallyEqual(test, a, b, include_named_exprs=True):
+    """unittest-based assertion for comparing expressions
+
+    This converts the expressions `a` and `b` into prefix notation and
+    then compares the resulting lists.  Operators and (non-native type)
+    leaf nodes in the prefix epresentation are converted to strings
```
Comment (emma58) [inline pyomo/core/expr/compare.py:230]: typo: representation

File: pyomo/core/expr/compare.py [hunk 2]
```diff
@@ -185,3 +189,79 @@ def compare_expressions(expr1, expr2, include_named_exprs=True):
...
+    # Convert leaf nodes and operators to their string equivalents
+    for prefix in (prefix_a, prefix_b):
+        for i, v in enumerate(prefix):
+            if type(v) in native_types:
+                continue
+            if type(v) is tuple:
+                if len(v) == 3:
+                    prefix[i] = v[:2] + (str(v[2]),)
+                continue
+            prefix[i] = str(v)
```
Comment (emma58) [inline pyomo/core/expr/compare.py:259]: I'm confused. Aren't some of these tuples with length 2? Like for named expressions? Maybe I'm just slow, but maybe you could comment what these cases are?

File: pyomo/core/expr/compare.py [hunk 3]
```diff
@@ -185,3 +189,79 @@ def compare_expressions(expr1, expr2, include_named_exprs=True):
...
+    try:
+        test.assertEqual(len(prefix_a), len(prefix_b))
+        for _a, _b in zip(prefix_a, prefix_b):
+            test.assertIs(_a.__class__, _b.__class__)
```
Comment (emma58) [inline pyomo/core/expr/compare.py:263]: Do you still need to check the class--aren't they all strings? Oh, wait, except for native types, right?

**Review by emma58 — APPROVED** [review]
This is exciting! :) I got a little lost in the cases for the converting-to-strings part of the structural equality, so maybe comments would be nice, but otherwise I think this looks good!

File: pyomo/core/expr/compare.py (hunk 3 above)
Comment (jsiirola) [reply pyomo/core/expr/compare.py:263]: Yes.  This is comparing elements of prefix, which may be native types, tuples, strings, etc.

File: pyomo/core/expr/compare.py (hunk 2 above)
Comment (jsiirola) [reply pyomo/core/expr/compare.py]: Added comments / explanation...

**Pull request merged**